[Layout][Grid] Mitigate incorrect baseline for vertical-lr flex/grid boxes.

yulunwu2 opened the pull request:
#### bc8615f6c27ba7b374ae8d88aa0497bff1f986cb
<pre>
[Layout][Grid] Mitigate incorrect baseline for vertical-lr flex/grid boxes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296443">https://bugs.webkit.org/show_bug.cgi?id=296443</a>
&lt;<a href="https://rdar.apple.com/156628380">rdar://156628380</a>&gt;

Reviewed by Alan Baradlay.

Inline layout currently performs an unnecessary coordinate flip in

InlineDisplayContentBuilder::mapInlineRectLogicalToVisual()

for vertical-lr content.

This is not necessary for boxes that provide the correct coordinates
and baselines, which are then broken by this flip.

This change introduces a temporary fix in

LayoutIntegrationGeometryUpdate::baselineForBox()

by performing another flip for line-inverted writing modes. This ensures
that the baselines computed by the LayoutIntegrationGeometryUpdate will
allow the content to be rendered correctly.

 * LayoutTests/TestExpectations:
 * LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-container-baseline-001-expected.txt:
 * Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
 (WebCore::LayoutIntegration::baselineForBox):

Canonical link: <a href="https://commits.webkit.org/298204@main">https://commits.webkit.org/298204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21c83322df99d7a2fa3948e9df6718d1107d89f1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114439 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34185 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120606 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65155 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c83dd3ff-047a-45bd-9079-3c3c8dd18faa) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116328 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34815 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42745 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86970 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41870 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8865b517-b326-429a-975d-bdb06639654c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117387 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27740 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102775 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67363 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26922 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20901 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64291 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97117 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21016 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123811 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41453 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30928 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95789 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41830 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98966 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95574 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/24379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40731 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18571 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37482 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41333 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46841 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40924 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44230 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42674 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2009,7 +2010,20 @@ std::optional<LayoutUnit> RenderGrid::firstLineBaseline() const
+            if (auto* blockFlow = dynamicDowncast<RenderBlockFlow>(baselineGridItem.get())) {
+                if (!blockFlow->hasNonVisibleOverflow() && blockFlow->childrenInline() && blockFlow->hasLines()) {
+                    if (CheckedPtr inlineLayout = blockFlow->inlineLayout())
+                        baseline = inlineLayout->lastLineLogicalBaseline();
```
Why `lastLineLogicalBaseline`? This works for `grid-inline-baseline.html` since it only has one line but does it work if there are multiple lines?

yulunwu2 replied:
Yup, this works for multiple lines. Using firstLineLogicalBaseline like this:

https://github.com/WebKit/WebKit/commit/8374f08c6cc346568b74f6ce55739378f6b0b6d3

breaks a ton of tests due to using incorrect baselines for multi-line render block flows.

yulunwu2 replied:
I think there was an issue with how I was getting the first line logical baseline that was causing the issue, uploading another PR for EWS

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2009,7 +2010,20 @@ std::optional<LayoutUnit> RenderGrid::firstLineBaseline() const
+        if (!GridLayoutFunctions::isOrthogonalParent(*this, *this->parent())) {
+            if (auto* blockFlow = dynamicDowncast<RenderBlockFlow>(baselineGridItem.get())) {
+                if (!blockFlow->hasNonVisibleOverflow() && blockFlow->childrenInline() && blockFlow->hasLines()) {
+                    if (CheckedPtr inlineLayout = blockFlow->inlineLayout())
```
In general we probably shouldn't try to introspect the grid item like this and hook into inline layout logic. Grids shouldn't care or have to worry about the contents of their grid items like this since the item itself establishes an independent formatting context. 


If you want to restrict your logic to apply to only certain types of boxes I think instead we should try to check what type of box it is (e.g. a block container) via its display type instead of type checking the renderer. 

If you need to get the logical baseline instead of the physical baseline (which is what it seems like `RenderBlockFlow` is doing) then maybe we should consider adding some sort of additional API that instead does it?

yulunwu2 replied:
I agree that hooking into inline layout logic here may not be the best thing to do. 

The LayoutIntegrationBoxGeometryUpdater calling into this function is regularly checking the renderer's lastLineLogicalBaseline()

https://searchfox.org/wubkat/source/Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp#425-426

https://searchfox.org/wubkat/source/Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp#547-548

I can refactor the logic out of RenderGrid and into 

LayoutIntegration::baselineForBox

here

https://searchfox.org/wubkat/source/Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp#468-472

if you prefer.

sammygill reviewed (commented):
I noticed that `RenderBlockFlow` is the only caller of `{first, last}LinePhysicalBaseline. Would it work if instead we changed it to use the logical variants? I haven't thought about it too much but from a cursory glance maybe it's ok? Might be worth considering!

yulunwu2 committed:
[Layout][Grid] Mitigate incorrect baseline for vertical-lr flex/grid boxes.
https://bugs.webkit.org/show_bug.cgi?id=296443
<rdar://156628380>

Reviewed by Alan Baradlay.

Inline layout currently performs an unnecessary coordinate flip in

InlineDisplayContentBuilder::mapInlineRectLogicalToVisual()

for vertical-lr content.

This is not necessary for boxes that provide the correct coordinates
and baselines, which are then broken by this flip.

This change introduces a temporary fix in

LayoutIntegrationGeometryUpdate::baselineForBox()

by performing another flip for line-inverted writing modes. This ensures
that the baselines computed by the LayoutIntegrationGeometryUpdate will
allow the content to be rendered correctly.

 * LayoutTests/TestExpectations:
 * LayoutTests/imported/w3c/web-platform-tests/css/css-grid/alignment/grid-container-baseline-001-expected.txt:
 * Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
 (WebCore::LayoutIntegration::baselineForBox):

Canonical link: https://commits.webkit.org/298204@main
